### PR TITLE
coq_rules: generate correct plugin dependency when ocamlopt not available

### LINKED ISSUE #18609: Coq fails to build on i686 with `-native-compiler no` due to `Error: No rule found for plugins/ssr/ssreflect_plugin.cmxs` (among similar others)
#### Description of the problem

I'm trying to build Coq for Void Linux, the i686 target. It fails with errors like

```
File "theories/dune", line 1807, characters 1-778:
1807 |  (rule
1808 |   (targets ssreflect.timing ssreflect.glob ssreflect.vos ssreflect.vo)
1809 |   (deps ../../theories/Init/Prelude.vo ../../theories/Bool/Bool.vo
1810 |         ../../theories/ssrmatching/ssrmatching.vo
1811 |         ../../plugins/ssr/ssreflect_plugin.cmxs
1812 |         ../../theories/ssr/ssrunder.vo)
1813 |   (action (chdir ../../. (run coqc -boot -R theories Coq -Q user-contrib/Ltac2 Ltac2 -I plugins/syntax -I plugins/funind -I plugins/micromega -I plugins/ltac -I plugins/extraction -I plugins/ssrmatching -I plugins/btauto -I plugins/nsatz -I plugins/ring -I plugins/rtauto -I plugins/derive -I plugins/firstorder -I plugins/ssr -I plugins/ltac2 -I plugins/cc -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file theories/ssr/ssreflect.timing %{dep:ssreflect.v}))))
Error: No rule found for plugins/ssr/ssreflect_plugin.cmxs
```

Here is the configuration

```
You have OCaml 5.1.0. Good!
You have OCamlfind 1.9.6. Good!
Warning: Cannot find the OCaml native-code compiler.
Only the bytecode version of Coq will be available.
You have the Zarith library 1.13 installed. Good!

  Architecture                : Linux
  Sys.os_type                 : Unix
  OCaml version               : 5.1.0
  OCaml binaries in           : /usr/bin/
  OCaml library in            : /usr/lib/ocaml
  Web browser                 : firefox -remote "OpenURL(%s,new-tab)" || firefox %s &
  Coq web site                : http://coq.inria.fr/
  Bytecode VM enabled         : true
  Native Compiler enabled     : no

  Paths where installation is expected by Coq Makefile:
  - Coq is expected in /usr
  - the Coq library is expected in /usr/lib/coq
  - the Coqide configuration files is expected in /usr/etc/xdg/coq
  - the Coqide data files is expected in /usr/share/coq
  - the Coq man pages is expected in /usr/share/man
  - documentation prefix path for all Coq packages is expected in /usr/share/doc
```

Can anyone help me with insights on this?

#### Coq Version

18.9

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #18609

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Not sure if going through coq_config to detect byte-only mode is the correct way to do it or if there's a better option.

**Comment by ejgallego** [comment]
> Not sure if going through coq_config to detect byte-only mode is the correct way to do it or if there's a better option.

That's a good question. I'd rather keep this stuff outside configure, unless we are willing to test it in the CI.

Instead I'd add an argument to `gen_rules` if someone wants to play with that, plus a bit of logic that checks if `gen_rules` has been called with `Sys.backend_type <> Native` for the cases where `ocamlopt` is not available, so things work out of the box.

File: tools/dune_rule_gen/coq_rules.ml [hunk 1]
```diff
@@ -79,7 +79,12 @@ let path_of_dep ~vo_ext dep =
   let open Coqdeplib.Dep_info in
   let file = match dep with
     | Dep.Require dep -> dep ^ vo_ext
-    | Dep.Ml (dep, _ext)-> dep ^ ".cmxs"
+    | Dep.Ml (dep, kind)->
+      let suff =
+        if Coq_config.best_compiler = "opt" then ".cmxs"
+        else Dep.byte_suff kind
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply tools/dune_rule_gen/coq_rules.ml]: you mean cma but sure

**Comment by SkySkimmer** [comment]
I don't really want to touch the argument parsing in gen_rules, it looks pretty adhoc

**Comment by ejgallego** [comment]
> I don't really want to touch the argument parsing in gen_rules, it looks pretty adhoc

We can rely on `Sys.backend_type` then, but I think it would make sense to use the `ext_plugin` Dune variable to do this properly (IMHO Coq's configure must not re-do OCaml detection, but reuse the one provided by Dune)

I'd expect the rule to become:
```
(rule
 (targets theories_dune)
 (deps
  (source_tree theories)
  (source_tree plugins))
 (action
  (with-stdout-to %{targets}
   (run tools/dune_rule_gen/gen_rules.exe Coq theories --ext_plugin=%{ext_plugin} %{env:COQ_DUNE_EXTRA_OPT=}))))
```
Argument parsing is ad-hoc, but IMHO that's not a big deal, as this tool is not user-facing.

We could for example require the plugin extension string to be the first argument always.

p.s: Dune `main` can finally include generated dune files via `(dynamic_include file)`, so we should be able to remove the dunestrap process at some point.

**Comment by SkySkimmer** [comment]
Not really sure what you want and I don't have more time for this.

**Comment by ejgallego** [comment]
I think it would be good not to use configure for this, IMHO we should not need to reconfigure Coq for this. We can either detect if native is available or add a parameter.

Other than that this patch is straightforward.

**Pull request closed without merging**